org.knime.scijava.commands/src/org/knime/scijava/commands/module/DefaultNodeModule.java

```diff
@@ -0,0 +1,206 @@
+package org.knime.scijava.commands.module;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.knime.core.data.DataCell;
+import org.knime.core.data.DataRow;
+import org.knime.core.data.DataType;
+import org.knime.core.node.ExecutionContext;
+import org.knime.scijava.commands.CellOutput;
+import org.knime.scijava.commands.MultiOutputListener;
+import org.knime.scijava.commands.converter.KNIMEConverterService;
+import org.scijava.Context;
+import org.scijava.module.Module;
+import org.scijava.module.ModuleInfo;
+import org.scijava.module.ModuleItem;
+import org.scijava.module.ModuleService;
+import org.scijava.module.process.ModulePreprocessor;
+import org.scijava.module.process.PreprocessorPlugin;
+import org.scijava.plugin.Parameter;
+import org.scijava.plugin.PluginService;
+
+/**
+ * Default implementation of {@link NodeModule}.
+ *
+ * @author Christian Dietz, University of Konstanz
+ */
+class DefaultNodeModule implements NodeModule {
+
+    @Parameter
+    private PluginService ps;
+
+    @Parameter
+    private KNIMEConverterService cs;
+
+    @Parameter
+    private ModuleService ms;
+
+    // Internally used variables
+    private final Module module;
+
+    private final Map<Integer, ModuleItem<?>> inputMapping;
+
+    private final Map<ModuleItem<?>, DataType> outputMapping;
+
+    private final NodeModuleOutputChangedListener outputListener;
+
+    /**
+     * Constructor.
+     *
+     * @param context
+     *            Scijava context.
+     * @param info
+     *            info this module is created for
+     * @param params
+     *            preresolved values for the parameters of the module
+     * @param inputMapping
+     *            mapping of input column index to input items of the module
+     * @param outputMapping
+     *            mapping of output items to types for the cells to for them
+     */
+    public DefaultNodeModule(final Context context, final ModuleInfo info,
+            final Map<String, Object> params,
+            final Map<Integer, ModuleItem<?>> inputMapping,
+            final Map<ModuleItem<?>, DataType> outputMapping) {
+        context.inject(this);
+
+        this.inputMapping = inputMapping;
+        this.outputMapping = outputMapping;
+        this.module = ms.createModule(info);
+
+        // Setting parameters
+        for (final Entry<String, Object> entry : params.entrySet()) {
+            module.setInput(entry.getKey(), entry.getValue());
+            module.resolveInput(entry.getKey());
+        }
+
+        // FIXME: do we need them all?
```

Comment:
depends how costly the preprocessors are. As we've no put the preprocessing logic outside the run logic it seems to be fine to take all preprocessors. However, there are preprocessors which are really useless in case of headless execution.